File: Client/src/main/java/com/rspsi/editor/model/WorldWindow.java
package com.rspsi.editor.model;

import java.util.Objects;

/**
 * World-space placement of a canonical document window.
 *
 * <p>{@link LocalTile} stays local to a document while {@link WorldTile}
 * represents absolute OSRS coordinates. This type is the only conversion
 * boundary between those spaces.</p>
 */
public record WorldWindow(int originX, int originY, int width, int length) {
    public WorldWindow {
        if (width <= 0 || length <= 0) {
            throw new IllegalArgumentException("World window dimensions must be positive");
        }
    }

    public boolean contains(LocalTile coordinate) {
        Objects.requireNonNull(coordinate, "coordinate");
        return coordinate.x() >= 0 && coordinate.x() < width
                && coordinate.y() >= 0 && coordinate.y() < length;
    }

    public boolean contains(WorldTile coordinate) {
        Objects.requireNonNull(coordinate, "coordinate");
        return coordinate.x() >= originX && coordinate.x() < originX + width
                && coordinate.y() >= originY && coordinate.y() < originY + length;
    }

    public WorldTile toWorld(LocalTile coordinate) {
        requireContains(coordinate);
        return new WorldTile(coordinate.plane(),
                originX + coordinate.x(), originY + coordinate.y());
    }

    public java.util.Optional<LocalTile> tryToLocal(WorldTile coordinate) {
        Objects.requireNonNull(coordinate, "coordinate");
        if (!contains(coordinate)) return java.util.Optional.empty();
        return java.util.Optional.of(new LocalTile(coordinate.plane(),
                coordinate.x() - originX, coordinate.y() - originY));
    }

    public LocalTile toLocal(WorldTile coordinate) {
        return tryToLocal(coordinate).orElseThrow(() ->
                new IndexOutOfBoundsException("World tile outside window: " + coordinate));
    }

    public int worldX(LocalTile coordinate) {
        requireContains(coordinate);
        return originX + coordinate.x();
    }

    public int worldY(LocalTile coordinate) {
        requireContains(coordinate);
        return originY + coordinate.y();
    }

    /**
     * Compatibility bridge for older local-coordinate code. New code should
     * use LocalTile explicitly.
     */
    @Deprecated
    public boolean contains(TileCoordinate coordinate) {
        return coordinate != null && contains(LocalTile.from(coordinate));
    }

    @Deprecated
    public int worldX(TileCoordinate coordinate) {
        return worldX(LocalTile.from(coordinate));
    }

    @Deprecated
    public int worldY(TileCoordinate coordinate) {
        return worldY(LocalTile.from(coordinate));
    }

    private void requireContains(LocalTile coordinate) {
        if (!contains(coordinate)) {
            throw new IndexOutOfBoundsException("Tile outside world window: " + coordinate);
        }
    }
}

File: Client/src/main/kotlin/com/rspsi/editor/model/WorldWindow.kt
package com.rspsi.editor.model

import java.util.Optional

/**
 * World-space placement of one canonical local document window.
 *
 * [LocalTile] remains document-local while [WorldTile] represents absolute OSRS
 * coordinates. This type is the explicit conversion boundary between those spaces;
 * callers should not reinterpret matching numeric components across the two types.
 */
@JvmRecord
data class WorldWindow(
    val originX: Int,
    val originY: Int,
    val width: Int,
    val length: Int,
) {
    init {
        if (width <= 0 || length <= 0) {
            throw IllegalArgumentException("World window dimensions must be positive")
        }
    }

    fun contains(coordinate: LocalTile?): Boolean {
        val safeCoordinate = coordinate ?: throw NullPointerException("coordinate")
        return safeCoordinate.x >= 0 &&
            safeCoordinate.x < width &&
            safeCoordinate.y >= 0 &&
            safeCoordinate.y < length
    }

    fun contains(coordinate: WorldTile?): Boolean {
        val safeCoordinate = coordinate ?: throw NullPointerException("coordinate")
        return safeCoordinate.x >= originX &&
            safeCoordinate.x < originX + width &&
            safeCoordinate.y >= originY &&
            safeCoordinate.y < originY + length
    }

    fun toWorld(coordinate: LocalTile?): WorldTile {
        val safeCoordinate = requireContains(coordinate)
        return WorldTile(
            safeCoordinate.plane,
            originX + safeCoordinate.x,
            originY + safeCoordinate.y,
        )
    }

    fun tryToLocal(coordinate: WorldTile?): Optional<LocalTile> {
        val safeCoordinate = coordinate ?: throw NullPointerException("coordinate")
        if (!contains(safeCoordinate)) {
            return Optional.empty()
        }
        return Optional.of(
            LocalTile(
                safeCoordinate.plane,
                safeCoordinate.x - originX,
                safeCoordinate.y - originY,
            ),
        )
    }

    fun toLocal(coordinate: WorldTile?): LocalTile =
        tryToLocal(coordinate).orElseThrow {
            IndexOutOfBoundsException("World tile outside window: $coordinate")
        }

    fun worldX(coordinate: LocalTile?): Int {
        val safeCoordinate = requireContains(coordinate)
        return originX + safeCoordinate.x
    }

    fun worldY(coordinate: LocalTile?): Int {
        val safeCoordinate = requireContains(coordinate)
        return originY + safeCoordinate.y
    }

    /**
     * Compatibility bridge for older local-coordinate code.
     *
     * New code should use [LocalTile] explicitly. This overload intentionally remains only
     * to support command/history/selection code that has not completed its coordinate migration.
     */
    @Deprecated("Use LocalTile explicitly")
    fun contains(coordinate: TileCoordinate?): Boolean =
        coordinate != null && contains(LocalTile.from(coordinate))

    @Deprecated("Use LocalTile explicitly")
    fun worldX(coordinate: TileCoordinate?): Int =
        worldX(LocalTile.from(coordinate))

    @Deprecated("Use LocalTile explicitly")
    fun worldY(coordinate: TileCoordinate?): Int =
        worldY(LocalTile.from(coordinate))

    private fun requireContains(coordinate: LocalTile?): LocalTile {
        val safeCoordinate = coordinate ?: throw NullPointerException("coordinate")
        if (!contains(safeCoordinate)) {
            throw IndexOutOfBoundsException("Tile outside world window: $safeCoordinate")
        }
        return safeCoordinate
    }

    override fun toString(): String =
        "WorldWindow[originX=$originX, originY=$originY, width=$width, length=$length]"
}
